### PR TITLE
feat(subscriptions): Use RPC in subscriptions pipeline

### PR DESCRIPTION
**Pull request by shruthilayaj**
Updates the subscription pipeline to support RPCSubscriptionData.

### CODE REVIEW

File: snuba/subscriptions/data.py [hunk 1]
```diff
@@ -280,6 +332,158 @@ def to_dict(self) -> Mapping[str, Any]:
...
+
+        request_class = EndpointTimeSeries().request_class()()
+        request_class.ParseFromString(base64.b64decode(self.time_series_request))
+
+        # TODO: update it to round to the lowest granularity
+        # rounded_ts = int(timestamp.replace(tzinfo=UTC).timestamp() / 15) * 15
+        rounded_ts = (
+            int(timestamp.replace(tzinfo=UTC).timestamp() / self.time_window_sec)
+            * self.time_window_sec
+        )
```
Comment (shruthilayaj) [inline snuba/subscriptions/data.py:373]: hack for now, we want to round this to the lowest available granularity (15s) see https://sentry.slack.com/archives/C07M2RN1AHJ/p1730738626703789

File: snuba/subscriptions/data.py [hunk 2]
```diff
@@ -280,6 +332,158 @@ def to_dict(self) -> Mapping[str, Any]:
...
+    request_version: str
+
+    def validate(self) -> None:
+        super().validate()
+        if (self.request_name, self.request_version) not in REQUEST_TYPE_ALLOWLIST:
+            raise InvalidSubscriptionError(
+                f"{self.request_name} {self.request_version} not supported."
+            )
+
+        # TODO: Validate no group by, having, order by etc
```
Comment (shruthilayaj) [inline snuba/subscriptions/data.py:353]: I'll add more validation in a follow up so it's easier to review

**Comment by shruthilayaj** [comment]
> Why can the create subscription RPC call not be its own PR?

@volokluev Creating a subscription requires validation, so need to be able to build and run the request before storing it. That introduces the RPCSubscriptionData class. And basically once I introduced it in one place, I kinda had to follow all the typing issues and handle it everywhere.

File: snuba/subscriptions/data.py [hunk 3]
```diff
@@ -95,7 +104,7 @@ def from_string(cls, value: str) -> SubscriptionIdentifier:
 
 
 @dataclass(frozen=True, kw_only=True)
-class SubscriptionData(ABC):
+class _SubscriptionData(ABC, Generic[TRequest]):
```
Comment (shruthilayaj) [inline snuba/subscriptions/data.py:107]: `build_request` returns generic type `TRequest` and `run_query` accepts `TRequest`.
So for SnQLSubscriptionData, `build_request` returns `Request` and `run_query` accepts `Request`
and for RPCSubscriptionData, `build_request` returns `TimeSeriesRequest` and `run_query` accepts `TimeSeriesRequest`

File: tests/subscriptions/test_filter_subscriptions.py [hunk 4]
```diff
@@ -60,3 +77,74 @@ def test_filter_subscriptions(expected_subs, extra_subs) -> None:  # type: ignor
...
+def test_filter_rpc_subscriptions(expected_rpc_subs, extra_rpc_subs) -> None:  # type: ignore
+    importlib.reload(scheduler)
+
+    filtered_subs = filter_subscriptions(
+        subscriptions=expected_rpc_subs + extra_rpc_subs,
+        entity_key=EntityKey.EAP_SPANS,
+        metrics=DummyMetricsBackend(strict=True),
+        slice_id=2,
+    )
+    assert filtered_subs == expected_rpc_subs
```
Comment (volokluev) [inline tests/subscriptions/test_filter_subscriptions.py:150]: this test is not necessary.

1. eap_spans is not a sliced storage
2. this is far in the weeds of the implementation details of the subscription scheduler

File: snuba/datasets/slicing.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 should be stored. These do not require individual physical partitions but allow
 for repartitioning with less code changes per physical change.
 """
+
```
Comment (volokluev) [inline snuba/datasets/slicing.py:6]: 🔪

File: snuba/subscriptions/codecs.py [hunk 6]
```diff
@@ -98,15 +116,20 @@ def decode(self, value: KafkaPayload) -> ScheduledSubscriptionTask:
 
         entity_key = EntityKey(scheduled_subscription_dict["entity"])
 
+        data = scheduled_subscription_dict["task"]["data"]
```
Comment (volokluev) [inline snuba/subscriptions/codecs.py:119]: FYI: the docstring of this class should probably change

**Review by volokluev — APPROVED** [review]
I have some cleanup comments but overall this is good to go

**Comment by volokluev** [comment]
When you go to merge this, please notify the oncall in `#discuss-eng-sns` this change is touching a lot of crucial infra. It's good you have done an end to end test locally. We should make sure that it doesn't break things in S4S when it's rolled out

**Comment by shruthilayaj** [comment]
> When you go to merge this, please notify the oncall in #discuss-eng-sns this change is touching a lot of crucial infra. It's good you have done an end to end test locally. We should make sure that it doesn't break things in S4S when it's rolled out

Will do, I'll address your comments and merge tomorrow 👍

File: snuba/subscriptions/codecs.py (hunk 6 above)
Comment (shruthilayaj) [reply snuba/subscriptions/codecs.py:119]: https://github.com/getsentry/snuba/pull/6499/commits/a642b1b997cf6697615f5406e79b94d5dbdfcf6a#diff-f48e31c677eccff7bf14b091fe348640c394684c53c8d3a5a16d27e265d8b1abR87

**Pull request merged**